File: qr.py
# -*- coding: utf-8 -*-

import numpy.linalg
import numpy as np
from time import time 

def wilkinson_shift(matrix):
    n = matrix.shape[0]
    if n == 1:
        return matrix[0, 0]
    
    # Extract the bottom-right 2x2 submatrix
    submatrix = matrix[-2:, -2:]
    
    # Compute the eigenvalues of the submatrix
    a, b, c, d = submatrix[0, 0], submatrix[0, 1], submatrix[1, 0], submatrix[1, 1]
    trace = a + d
    det = a * d - b * c
    #Quadratic Formula
    mu1 = trace / 2 + np.sqrt((trace / 2)**2 - det)
    mu2 = trace / 2 - np.sqrt((trace / 2)**2 - det)
    
    # Choose the eigenvalue that is closer to the bottom-right element
    if abs(d - mu1) < abs(d - mu2):
        return mu1
    else:
        return mu2
    
    #computes norm of a vector
def vec_norm(v):
    return np.sqrt(np.sum(np.power(v,2)))

def v_vec_norm(v):
    if (v.shape[0] > 1):
        return np.sqrt(np.power(v[0], 2) + np.power(v[1], 2))
    else:
        return np.sqrt(np.power(v[0],2))
    
   
def v_vec_dot(v):
    if (v.shape[0] > 1):
        return np.power(v[0], 2) + np.power(v[1], 2)
    else:
        return v[0]*v[0]
    
    #helps to converge faster, zeros out some elements
def reflection_vector(x):
    v = np.copy(x)
    v[0] += np.sign(x[0]) *v_vec_norm(x)
    eps = 1e-10  # Small epsilon to avoid division by zero
    return v, 2.0 / (v_vec_dot(v) + eps)

def qr_householder(A):
    n = A.shape[0]
    Q = np.eye(n)
    R = np.copy(A)
    
    for j in range(n-1):
        v, c = reflection_vector(R[j:, j])
        
        # Restrict v to the first 2 elements
        v = v[:2]
        
        # Update R
        R_sub = R[j:j+2, j:]
        R[j:j+2, j:] -= c * np.outer(v, v @ R_sub)
        
        # Update Q
        Q_sub = Q[:, j:j+2]
        Q[:, j:j+2] -= c * np.outer(Q_sub @ v, v)

    return Q, R

     
def hessenberg_form(A, epsilon=1e-6):
    
    n = A.shape[0]
    H = np.copy(A)
    Q = np.eye(n)

    for k in range(n - 2):
        x = H[k+1:, k]#extracts the elements in the k-th column under the main diagonal
        v = np.copy(x)
        v[0] += np.sign(x[0]) * vec_norm(x)
        vNorm=vec_norm(v)
        if(vNorm!=0):#added conditon to avoid division by zero
         v /= vec_norm(v)#normalize
        H[k+1:, k:] -= 2.0 * np.outer(v, v @ H[k+1:, k:])#update the lower submatrix starting from (k+1,k), an index under main diag
        H[:, k+1:] -= 2.0 * np.outer(H[:, k+1:] @ v, v)#update columns starting for the k+1 col
        Q[:, k+1:] -= 2.0 * np.outer(Q[:, k+1:] @ v, v)

    # Create a mask for  elements near the diagonal, assign true to the main diagonal and the diagonals above and below it
    mask = np.eye(n, dtype=bool) | np.eye(n, k=1, dtype=bool) | np.eye(n, k=-1, dtype=bool)


    #zeros elements with a false mask value that their absolute values are smaller than epsilon
    H[~mask] = np.where(np.abs(H[~mask]) < epsilon, 0, H[~mask])
    
    return H, Q

def myEigenRecursive(A, epsilon=1e-6):
    n = A.shape[0]
    I = np.eye(n)   
    Q = np.eye(n) 
    
    if n == 1: 
        return A[:, 0], I

    eigenvectors = np.eye(n)

    while np.min(np.abs(np.diag(A, 1))) > epsilon:#continue while the min elemnet abs value in the upper diagonal is bigger than epsilon 
        u = wilkinson_shift(A)# Use Wilkinson's shift
        uI = u * I
        Q, R = qr_householder(A - uI)# Apply QR factorization on the matrix A - uI
        A = R @ Q + uI                  
        eigenvectors = eigenvectors @ Q # update eigenvectors for the final result

    diag_arr_position = np.argmin(np.abs(np.diag(A, 1)))    

    upper_mat = A[:diag_arr_position + 1, :diag_arr_position + 1]
    low_mat = A[diag_arr_position + 1:, diag_arr_position + 1:]

    eigenvalues_upper, eigenvector_upper = myEigenRecursive(upper_mat, epsilon)
    eigenvalues_lower, eigenvector_lower = myEigenRecursive(low_mat, epsilon)
    
    eigenvalues = np.concatenate([eigenvalues_upper, eigenvalues_lower])#concat the eigenvalues returned from the two calls
    
    v1 = np.r_[eigenvector_upper, np.zeros((eigenvector_lower.shape[0], eigenvector_upper.shape[1]))]#vertical concat, padds eigenvector upper with 0's'
    v2 = np.r_[np.zeros((eigenvector_upper.shape[0], eigenvector_lower.shape[1])), eigenvector_lower]
    
    return eigenvalues, eigenvectors @ np.c_[v1, v2]#concat v1 v2 horizontally



def sort_by_same_order(eigenvalues, eigenvectors):
    sorted_indices = np.argsort(eigenvalues)
    sorted_eigenvalues = eigenvalues[sorted_indices]
    sorted_eigenvectors = eigenvectors[:, sorted_indices]
    return sorted_eigenvalues, sorted_eigenvectors


def is_symmetric(matrix):
    matrix = np.array(matrix)
    return np.array_equal(matrix, matrix.T)

if __name__ == '__main__':
    matrix = np.loadtxt("inv_matrix(800 x 800).txt", dtype='f', delimiter=' ')
    print(is_symmetric(matrix))
    print("Converting matrix to Hessenberg form...")
    tt = time()
    hessen_mat, H = hessenberg_form(matrix,1e-6)
    print("Calculating eigenvalues & eigenvectors of matrix using QR Algorithm...") 
    values, vectors = myEigenRecursive(hessen_mat, epsilon=1e-6)
    vectors = H @ vectors
    duration=time()-tt
    print("\nmy_eigen_recursive:\t", duration)
    tt = time()
    eigenvalues, eigenvectors = np.linalg.eig(matrix)
    duration=time()-tt
    print("np.linalg.eig: \t\t", duration)
    sorted_values, sorted_vectors = sort_by_same_order(values, vectors)
    sorted_eigenvalues, sorted_eigenvectors = sort_by_same_order(eigenvalues, eigenvectors)
    np.savetxt('sorted_vectors_python.txt', (np.abs(sorted_vectors)-np.abs(sorted_eigenvectors)), fmt='%.8f', delimiter=' ', comments='') 
    np.savetxt('sorted_values_python.txt', (np.abs(sorted_values)-np.abs(sorted_eigenvalues)), fmt='%.8f', delimiter=' ', comments='')
    print("\nEigenvector norm difference : \t", np.linalg.norm(np.abs(sorted_vectors) - np.abs(sorted_eigenvectors)))
    print("Eigenvalue norm difference  : \t", np.linalg.norm(np.abs(sorted_values) - np.abs(sorted_eigenvalues))) 

File: qr.hs
{-# LANGUAGE BangPatterns #-}


import Control.Monad
import Control.Monad.ST
import Data.List (sortOn)
import Data.Time.Clock
import Numeric.LinearAlgebra
import Numeric.LinearAlgebra.Devel
import Prelude hiding ((<>))

type MD = Matrix Double

type VD = Vector Double

sortBySameOrder :: VD -> MD -> (VD, MD)
sortBySameOrder eigenvalues eigenvectors = 
    let indices = [0 .. size eigenvalues - 1]
        sortedIndices = map snd $ sortOn fst $ zip (toList eigenvalues) indices
        sortedEigenvalues = fromList [eigenvalues `atIndex` i | i <- sortedIndices]
        sortedEigenvectors = eigenvectors ¿ sortedIndices
    in (sortedEigenvalues, sortedEigenvectors)

wilkinson_shift :: Matrix Double -> (Double)
wilkinson_shift !mat =
  let n = rows mat
   in if n == 1
        then mat `atIndex` (0, 0)
        else
          let subMat = subMatrix (n - 2, n - 2) (2, 2) mat
              a = subMat `atIndex` (0, 0)
              b = subMat `atIndex` (0, 1)
              c = subMat `atIndex` (1, 0)
              d = subMat `atIndex` (1, 1)
              trace = a + d
              det = a * d - b * c
              mul1 = trace / 2 + sqrt ((trace / 2) ** 2 - det)
              mul2 = trace / 2 - sqrt ((trace / 2) ** 2 - det)
           in if abs (d - mul1) < abs (d - mul2)
                then mul1
                else mul2




sign :: Double -> Double
sign x
  | x > 0 = 1.0
  | x < 0 = -1.0
  | otherwise = 0.0

vecNorm :: VD -> Double
vecNorm v = sqrt $ sumElements $ cmap (^ 2) v


vec_dot :: VD -> Double
vec_dot !x = c where
  vsize = size x
  a = at' x 0
  b = if vsize > 1 then (at' x 1) else 0.0
  c=if vsize > 1
    then  (a * a + b * b)
    else  (a * a)

vec_norm :: VD -> Double
vec_norm x = n where
  vsize = size x
  a = at' x 0
  b = if vsize > 1 then (at' x 1) else 0.0
  n=if vsize > 1
     then  (sqrt (a * a + b * b))
     else  (sqrt (a * a))

reflection_vector :: VD -> (VD, Double)
reflection_vector x = (vec, c)
  where
    x0=at' x  0
    v0=x0+(sign x0)*vec_norm x
    v=vjoin[vector [v0],subVector 1 ((size x) -1) x]
    eps=1e-10
    vec=v
    c= 2.0/(vec_dot(v)+eps)






qr_factorization_householder :: MD -> (MD, MD)
qr_factorization_householder !a = (q, r)
  where
    !n = rows a
    (q, r) = runST $ do
      r' <- thawMatrix a :: ST s (STMatrix s Double)
      q' <- thawMatrix (ident n) :: ST s (STMatrix s Double)
      forM_ [0 .. n - 2] $ \j -> do
        temp <- unsafeFreezeMatrix r'
        let subR = subMatrix (j, j) (n - j, 1) temp
        let (v, c) = reflection_vector $ flatten subR
        let v'=subVector 0 2 v
        let subR = subMatrix (j, j) (2, n - j) temp
        let newR = subR-(scale c (v' `outer` (v'<# subR)))
        setMatrix r' j j newR

        temp2 <- unsafeFreezeMatrix q'
        let subQ = subMatrix (0, j) (n, 2) temp2
        let newQ = subQ-(scale c ((subQ#>v') `outer` v'))
        setMatrix q' 0 j newQ

      changedQ <- unsafeFreezeMatrix q'
      changedR <- unsafeFreezeMatrix r'
      return (changedQ, changedR)

-----------Matrix A         eigenvectors  epsilon         return  MatA       eigenvectors
runLoop :: Matrix Double -> Matrix Double ->Double-> (Matrix Double, Matrix Double)
runLoop a eigenvectors epsilon = go a eigenvectors where
  go !mat !vectors |(((minElement $ cmap abs $ takeDiag $ subMatrix (0, 1) ((((rows mat) - 1)), ((rows mat) - 1)) mat)) <= epsilon) =(mat,vectors)|otherwise= 
    go newMat newVec where 
       u = wilkinson_shift mat
       ui = scale u (ident (rows mat))
       (q, r) = qr_factorization_householder (mat - ui)
       newMat = ((r Numeric.LinearAlgebra.<> q) + ui)
       newVec = vectors Numeric.LinearAlgebra.<> q


myEigenRecursive :: Matrix Double -> Double -> (Vector Double, Matrix Double)
myEigenRecursive a !epsilon =
    if rows a == 1
      then (Numeric.LinearAlgebra.fromList [a `atIndex` (0, 0)], ident 1)
      else
                      (eigenvalues, eigenvectors Numeric.LinearAlgebra.<> (v1 ||| v2))
  where
    n = rows a
    q = ident n

    (finalA, eigenvectors) = runLoop a q epsilon

    diagArrPosition = minIndex $ cmap abs $ takeDiag $ subMatrix (0, 1) (n - 1, n - 1) finalA

    upperMat = subMatrix (0, 0) (diagArrPosition + 1, diagArrPosition + 1) finalA
    lowMat = subMatrix (diagArrPosition + 1, diagArrPosition + 1) (n - diagArrPosition - 1, n - diagArrPosition - 1) finalA

    (eigenvaluesUpper, eigenvectorUpper) = myEigenRecursive upperMat epsilon
    (eigenvaluesLower, eigenvectorLower) = myEigenRecursive lowMat epsilon

    eigenvalues = vjoin [eigenvaluesUpper, eigenvaluesLower]

    v1 = eigenvectorUpper === konst 0.0 (rows eigenvectorLower, cols eigenvectorUpper)
    v2 = konst 0.0 (rows eigenvectorUpper, cols eigenvectorLower) === eigenvectorLower



createV :: VD -> VD
createV x = runST $ do
  v <- thawVector x :: ST s (STVector s Double)
  let first = at' x 0
  let res = first + (sign first) * (vecNorm x)
  writeVector v 0 res
  v' <- unsafeFreezeVector v
  let vNorm = vecNorm v'
  let n = if vNorm == 0 then 1.0 else vNorm
  return $ scale (1 / n) v'

applyMask :: [[Double]] -> Double -> [[Double]]
applyMask h epsilon =
  let mask i j = abs (i - j) <= 1
      applyMaskToElement i j x = if mask i j then x else if abs x < epsilon then 0 else x
   in [[applyMaskToElement i j x | (j, x) <- zip [0 ..] row] | (i, row) <- zip [0 ..] h]

hessen :: MD -> Double -> (MD, MD)
hessen a epsilon = (h, q)
  where
    n = rows a
    (h, q) = runST $ do
      h' <- thawMatrix a :: ST s (STMatrix s Double)
      q' <- thawMatrix (ident n) :: ST s (STMatrix s Double)
      forM_ [0 .. (n - 3)] $ \i -> do
        temp <- unsafeFreezeMatrix h'
        let b = subMatrix (i + 1, i) (n - i - 1, 1) temp
        let v = createV $ flatten b
        c <- unsafeFreezeMatrix h'
        let d = subMatrix (i + 1, i) (n - i - 1, n - i) c
        let vProduct = v <# d
        let out = d - 2.0 * (v `outer` vProduct)
        setMatrix h' (i + 1) i out
        e <- unsafeFreezeMatrix h'
        let m = subMatrix (0, i + 1) (n, n - i - 1) e
        let v2 = m #> v
        let out2 = m - 2.0 * (v2 `outer` v)
        setMatrix h' 0 (i + 1) out2
        l <- unsafeFreezeMatrix q'
        let o = subMatrix (0, i + 1) (n, n - i - 1) l
        let v3 = o #> v
        let out3 = o - 2.0 * (v3 `outer` v)
        setMatrix q' 0 (i + 1) out3
      f <- unsafeFreezeMatrix h'
      let aList = toLists f
      let newH = fromLists $ applyMask aList epsilon
      g <- unsafeFreezeMatrix q'
      return (newH, g)
      
main = do
  w<-loadMatrix "inv_matrix(800 x 800).txt"
  tt1 <- getCurrentTime
  let (!g, !k) = hessen w 1e-6
  let (a, j) = myEigenRecursive g 1e-6
  let vectors=k<>j
  disp 2 (subMatrix (0,0) (1,1) vectors)
  tt2 <- getCurrentTime
  print "my eig recursive:"
  print (diffUTCTime tt2 tt1)
  let (sorted_values, sorted_vectors)= sortBySameOrder a vectors
  tt1 <- getCurrentTime
  let (eigVal,eigVectors)=eig w
  print (subMatrix (0,0) (1,1) eigVectors)  
  tt2 <- getCurrentTime
  print "Hmatrix eig:"
  print (diffUTCTime tt2 tt1)
  let realEigenvectors = cmap realPart eigVectors
  let realEigenvalues = cmap realPart eigVal
  let (sortEigVal,sortEigVectors)=sortBySameOrder realEigenvalues realEigenvectors
  let absSortedVectors = cmap abs sorted_vectors
  let absSortedEigenvectors = cmap abs sortEigVectors
  let absSortedValues = cmap abs sorted_values
  let absSortedEigenvalues = cmap abs sortEigVal

-- Compute the differences
  let vectorDiff = absSortedVectors - absSortedEigenvectors
  let valueDiff = absSortedValues - absSortedEigenvalues

-- Compute the norms of the differences
  let vectorNormDifference = norm_2 vectorDiff
  let valueNormDifference = norm_2 valueDiff
  saveMatrix "vectorDiff_Haskell.txt" "%.8f" vectorDiff
  saveMatrix "valueDiff_Haskell.txt" "%.8f" (asColumn valueDiff)
-- Print the results
  putStrLn $ "\nEigenvector norm difference : \t" ++ show vectorNormDifference
  putStrLn $ "Eigenvalue norm difference  : \t" ++ show valueNormDifference


